### DISCUSSION TITLE
Customizable filenames

### DISCUSSION THREAD

**Discussion by Kyther** [nianeyna/ao3downloader#94] [Ideas]
Could there be a way to customize the downloaded filenames?  I have a specific filename pattern I use for every fanfic, so it'll take me a lot of time fixing them if they're not customizable.  I'm referring to having a pattern like %t for title, %a for author, etc., so that a user can set "%a - %t" to get "Author - Title" and "%t by %a" to get "Title by Author" and suchlike.  (For another example of how it could look, see fichub-cli, which implements customizable filenames through a config file and [author], [fandom], etc.  They don't offer authentication, however, and don't use the built-in download options from AO3, which makes it useful for ff.n, but I'm still looking for a better AO3 option and this one ticks all the boxes except for the filename issue.)

I saw some reference to putting the work id on every fic, which, for someone who dumps everything into the same folder or lets Calibre manage it or something, I can see how that might be needed.  However, I never pull mixed fandoms because I manually sort fics by fandom (I've got a few bash scripts to do the initial sorting within a fandom for me), so that isn't an issue, and I've not yet found an author writing two fics with the same title in the same fandom.  Removing work ids from every filename would take me ages!  Having the option to customize the filenames (even if it came with the caveat of "you might end up with the same filename!" warnings) would be amazing.

**Comment by nianeyna** [comment]
Ha, I promise you, authors posting multiple fics in the same fandom with the same title *definitely* happens 😂 I can't fathom *why* it happens, but it does! The work id in filenames was actually added specifically to counter that issue, because it came up far more than you would possibly expect.

That said, you are not the first to request configurable filenames, and while it is not yet possible it is indeed a planned feature! Please track issue #72 if you want to know when it goes live. 😃 I will take your feedback into consideration when deciding what issues to work on next!

**Reply by nianeyna** [reply]
btw, I would recommend *not* waiting until this feature is live before using ao3downloader. Works disappear from ao3 every day. What I would do is back up your stuff, but don't bother organizing it yet. Just leave it dumped in a folder somewhere. Then, when I get this feature out, use the "redownload" option on your folder of unorganized fics, and it will save everything for you with your preferred filenames (albeit at the cost of... well... redownloading everything). Any works that were deleted or otherwise lost since you did the original backup will be recorded in the log file, and you can easily grab those (hopefully) few fics out of the original download folder and rename them manually. I hope that helps!

**Reply by Kyther** [reply]
I'm already trying it. :)  Had to do it in a VM though as I'm working on mostly older systems, and it *would not work* unless I ran in the right Python (it did not like 3.8.10, which was the default on the VM! - at least 3.9.5 was also in the repo) and a more recent OS.  Ah well!  (I tend to run a lot of other scripts on a really old box that has Python 3.8 via miniconda, but if it calls libraries that box doesn't have - like GLIBC 2.14 which this required - I have to move over to the VM.)

Having to keep it in that folder will be an issue, though, as I don't keep a lot in the VM and it's a bit awkward to get at the files.  For now I suppose I'll just do "back it all up but don't do anything with it", so it can all just sit there out of the way.  Fics are, at least, not terribly large, so I won't run out of space in the VM too soon.

I subscribed! :D

Possibly relevant:  With the Fichub stuff I pretty much just download it all using the naming pattern and then my scripts (simple bash ones, lots of if/then/else and while loops, lol) create the author folders and move files in from the sorting folder (which is parallel to the fandom one).  They're set to alert me if a file has the same name as one already in the folders, and skip over it so I can decide what to do with it manually.  Typically I add a (1) or whatnot to the end of the filename if it's distinct and I want to keep it (which I've done with edited versions of stories).  Maybe have a way to choose if to automatically rewrite over or to append a distinguishing string to the end?  That way people could choose to use a naming pattern that didn't have work ids distinguishing them, but still ensure they didn't lose data due to overwriting.

The other thing I noticed was the file names were cut off - couldn't see full title names or anything (I'm guessing it was named "[workid] whatever AO3 provides" without any editing, and AO3 has a short limit to filenames that I'm always having to bypass by copying the title or author before I click on the download buttons, and then paste over the default name).  I presume if the filenames get customizable that problem would be eliminated since the full title/author/etc. would be pulled from AO3 and not the limited (no commas or other special characters) version that it does?  Of course, then there's the issue that Windows users have a more limited set of characters they can have in their filenames, but I'm used to substituting -- for a : and space, and _ for a question mark.

**Reply by nianeyna** [reply]
Yeah, 3.8 is just a bit too old to work with ao3downloader unfortunately. You don't have to necessarily keep the fics in any specific folder, though - the script should be able to redownload from any folder that the executing user can access. Which may still mean keeping them on the VM in general, but they don't have to stay in the specific "downloads" folder inside the script directory in order for redownloading to work. Actually it's kind of better to *not* target the downloads folder itself with redownload, that might cause a loop now that I think of it.

**Reply by Kyther** [reply]
Yeah, I tried setting up a conda environment with 3.9 - which worked fine - but the glibc requirement was a hard no-go.

Ah, good to know - it's easier if I stick them in fandom-specific subfolders anyway so that works.  I'm grabbing fandoms in separate batches by default.

**Reply by nianeyna** [reply]
Actually the *real* issue with Windows filenames isn't a limited character *set* it's a limited character *length*. Basically in order for the script to not completely crash on most people's computers I have to be *very* strict about the file name length - currently I only allow 50 characters (which is why you're noticing it get cut off). When I had it set longer *lots* of people were getting bugs. It kind of sucks for mac and linux users, I know, because it feels like I'm punishing you for no reason haha. Since you seem to know what you're doing though, [here](https://github.com/nianeyna/ao3downloader/blob/2b5acf90f011e4109c2a292eca8ffcd8103b659e/ao3downloader/parse_text.py#L20) is the relevant line of code which you can edit in your copy if you wish. You can either change it to a higher number, or remove `[:50]` entirely for an unlimited length. I might include an "unlimited file name length" option when I add the configurable file names, but I'd have to be careful with it because sadly I think some Windows users might select it thinking "oh that sounds good" and then not realize why all their downloads are failing.

Ultimately though... encoding a lot of metadata in file names is just kind of not a great solution. :/ Unfortunately, I don't know what's better.

For reference, you can see the full list of characters I don't allow in filenames [here](https://github.com/nianeyna/ao3downloader/blob/2b5acf90f011e4109c2a292eca8ffcd8103b659e/ao3downloader/strings.py#L4). As you can see, it's not long. The actual problem is [this](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry).

**Reply by Kyther** [reply]
Ah right, I forgot about Windows filename length restrictions.  Well, I forgot they were that bad, lol.  I'm so glad I'm on Linux…  I don't miss the days of renaming files to remove all spaces so I could still get the full title on the file and not have it complain at me.

I'll probably edit that later - right now the name format is so off what I use anyway that there's no point in trying to get it formatted the way I want until I can set it before it gets downloaded, and AO3 naturally cuts off the filename and strips out even commas and all that anyhow.  Yeah, I can see that if you wanted to offer that you'd have to have some big warning "DO NOT CHOOSE THIS IF YOU ARE ON WINDOWS" or something like that.  Even then, it's guaranteed that someone will choose it and not realize, but at least you could say you tried? lol

Yeah, the format I use is always "[title] by [author]" - I started it years ago, have a curated collection that's quite large already using that format (including stuff long-deleted from the sites they were on), and figure I'd better keep it up.  As far as keeping *track* of it all, well, I've another method entirely but I'm in the wrong account (and probably the wrong place) for discussing that. *g*

**Reply by verotheelf** [reply]
You can change the title format by editing this [code](https://github.com/nianeyna/ao3downloader/blob/2b5acf90f011e4109c2a292eca8ffcd8103b659e/ao3downloader/parse_soup.py#L183C4-L183C43)

**Reply by Kyther** [reply]
LOL, I hadn't even thought about it.  But I might just do that.  Make it what I would use.  Only thing is, if there *does* turn out to be two fics with identical titles and authors in the same fandom… what would it do?  I'd love if it just added a disambig such as " (1)" to the end, but that would require extra code to sort out, I bet.

EDIT:  I went ahead and did it.  I'll just have to make sure with larger fandoms (which will have a greater risk), that I either monitor and move out the fics right away so I can do the check & merge (or write something that'll do that for me), or split into smaller batches at a time so there's less chance of stuff getting overwritten or clashing, since I'm still not sure what it'll do if it finds the fic already existing under that name.

**Reply by verotheelf** [reply]
Not the cleanest, but minimized the amount of code editing required. Insert it between line [27-28](https://github.com/nianeyna/ao3downloader/blob/2b5acf90f011e4109c2a292eca8ffcd8103b659e/ao3downloader/fileio.py#L27)
```
        if os.path.exists(file):
            filename = os.path.splitext(filename)[0] + "_" + datetime.datetime.now().strftime('%m%d%Y%H%M%S%f') + os.path.splitext(filename)[1]
            file = os.path.join(self.downloadfolder, filename)
```

This assumes you aren't using windows with its restrictions as the timestamp is added after the filename is truncated for safety. I think I saw you're on Linux so shouldn't be a problem. If you remove the code that shortens the filenames you probably won't run into this issue very often anyway. Every time I've noticed duplicate titles by the same author it's because the title is something like blah blah rated t version and blah blah rated m version. When it gets shortened they become identical.

**Reply by Kyther** [reply]
Sweeeeeet!  Thank you!  Yeah, that makes sense why someone would have duplicate titles - because they have two versions of the same work.  And yes, I'm on Linux - which doesn't have the same issues with running into filename length limits so easily.  I know there's a limit eventually - but I haven't run into it in the course of doing anything as of yet, and I *probably* am not going to run into it here, even though I've removed the truncation code. :)

This is working just beautifully, I love it so much. :D

**Reply by Kyther** [reply]
@verotheelf Is there a way to tweak the code to put the timestamp at the very beginning of the filename instead of the end, or would that be too difficult?  The sorting script I set up looks at the author name (which I have at the end of the filename) to sort them into folders, and the way I have the regex set up means adding a timestamp at the end of the filename keeps the file from being sorted where it belongs.  At the beginning of the filename wouldn't cause that issue at all.  I've been trying to move everything out of the download folder as soon as a bunch of it gets in there, in case, and once I did catch a set of dupes and renamed it the way I usually would, but going forward it would be easier not to worry about that.

**Reply by verotheelf** [reply]
No, not difficult, just requires a minor tweak.

        if os.path.exists(file):
            file = os.path.join(self.downloadfolder, datetime.datetime.now().strftime('%m%d%Y%H%M%S%f') + filename)

I can't actually test it out right now because nothing is downloading for me, but let me know if it works.

**Reply by verotheelf** [reply]
Forgot to add a symbol after the the time stamp

```
   if os.path.exists(file):
        file = os.path.join(self.downloadfolder, datetime.datetime.now().strftime('%m%d%Y%H%M%S%f') + "_" + filename)
```

**Reply by Kyther** [reply]
I think it works, lol.  Last night it errored out only halfway through a batch of fics, went from 299 to "done" in a flash and I was like "suuuure", so I looked, and sure enough, it had downloaded exactly the formats for 299 and no more.  I went into the text file and removed the first 299 links then, and started it back up - but I forgot to move the "images" folder out.  It created a new images folder with the datetime string in front of it!  So I'm guessing that's evidence that this works. XD
